capsules/src/analog_comparator.rs

```diff
@@ -0,0 +1,156 @@
+//! Provides userspace access to the analog comparators on a board.
+//!
+//! Usage
+//! -----
+//!
+//! ```
+//! let ac_channels = static_init!(
+//!     [&'static sam4l::acifc::AcChannel; 2],
+//!     [
+//!         &sam4l::acifc::CHANNEL_AC0,
+//!         &sam4l::acifc::CHANNEL_AC1,
+//!     ]
+//! );
+//! let analog_comparator = static_init!(
+//!     capsules::analog_comparator::AnalogComparator<'static, sam4l::acifc::Acifc>,
+//!     capsules::analog_comparator::AnalogComparator::new(&mut sam4l::acifc::ACIFC, ac_channels)
+//! );
+//! sam4l::acifc::ACIFC.set_client(analog_comparator);
+//! ```
+//!
+//! ## Number of Analog Comparators
+//! The number of analog comparators available depends on the microcontroller/board used.
+//!
+//! ## Normal or Interrupt-based Comparison
+//! For a normal comparison or an interrupt-based comparison, just one analog
+//! comparator is necessary.
+//!
+//! For more information on how this capsule works, please take a look at the
+//! README: 00007_analog_comparator.md in doc/syscalls.
+
+// Author: Danilo Verhaert <verhaert@cs.stanford.edu>
+// Last modified August 7th, 2018
+
+/// Syscall driver number.
+pub const DRIVER_NUM: usize = 0x00007;
+
+use core::cell::Cell;
+use kernel::hil;
+use kernel::{AppId, Callback, Driver, ReturnCode};
+
+pub struct AnalogComparator<'a, A: hil::analog_comparator::AnalogComparator + 'a> {
+    analog_comparator: &'a A,
+    callback: Cell<Option<Callback>>,
+    channels: &'a [&'a <A as hil::analog_comparator::AnalogComparator>::Channel],
+}
+
+impl<'a, A: hil::analog_comparator::AnalogComparator> AnalogComparator<'a, A> {
+    pub fn new(
+        analog_comparator: &'a A,
+        channels: &'a [&'a <A as hil::analog_comparator::AnalogComparator>::Channel],
+    ) -> AnalogComparator<'a, A> {
+        AnalogComparator {
+            analog_comparator: analog_comparator,
+            channels: channels,
+            callback: Cell::new(None),
+        }
+    }
+
+    // Do a single comparison on a channel
+    fn comparison(&self, channel: usize) -> ReturnCode {
+        if channel >= self.channels.len() {
+            panic!("Please select a channel which exists on the current board/chip");
+        }
+        // Convert channel index
+        let chan = self.channels[channel];
+        let result = self.analog_comparator.comparison(chan);
+
+        return ReturnCode::SuccessWithValue {
+            value: result as usize,
+        };
+    }
+
+    // Start comparing on a channel
+    fn start_comparing(&self, channel: usize) -> ReturnCode {
+        if channel >= self.channels.len() {
+            panic!("Please select a channel which exists on the current board/chip");
```

Comment:
We shouldn't have `panic!()` in capsules. Change this to a return code that signals the error to the app.